### PR TITLE
Better SDPA unmasking implementation

### PR DESCRIPTION
**Pull request by fxmarty**
As @ArthurZucker improved the unmasking for SDPA for mem-efficient code path let's do so for all archs using SDPA https://github.com/huggingface/transformers/pull/27931

### CODE REVIEW

**Comment by fxmarty** [comment]
`RUN_SLOW=1 CUDA_VISIBLE_DEVICES=3 pytest tests/ -k "test_eager_matches_sdpa_inference" -s -vvvvv` passes except for qwen2 (but it is unrelated, see https://github.com/huggingface/transformers/pull/28436#issuecomment-1966626609)

File: src/transformers/modeling_attn_mask_utils.py [hunk 1]
```diff
@@ -199,80 +200,11 @@ def _unmask_unattended(
...
-                    torch.arange(num_masks)[None, :, None],
-                    range_tensor[:, None, :],
-                )
-        else:
-            # Broadcast [left_masked_rows, 1], [left_masked_rows, max_len]
-            mask_slice = (left_masked_rows[:, None], range_tensor)
-
-        expanded_mask[mask_slice] = unmasked_value
+        if expanded_mask.dtype == torch.bool:
+            raise ValueError("AttentionMaskConverter._unmask_unattended expects a float `expanded_mask`, got a BoolTensor.")
```
Comment (fxmarty) [inline src/transformers/modeling_attn_mask_utils.py:205]: Some models (gpt bigcode) use bool tensors, but Arthur's implem can't work for that dtype.

Comment (ArthurZucker) [reply src/transformers/modeling_attn_mask_utils.py:205]: can't we cast it and replace the `min` with 0?

File: src/transformers/modeling_attn_mask_utils.py [hunk 2]
```diff
@@ -199,80 +200,11 @@ def _unmask_unattended(
         `attention_mask` is [bsz, src_seq_len].
 
         The dimension num_masks of `expanded_mask` is most often 1, but it can also be the number of heads in the case of alibi attention bias.
-
```
Comment (ArthurZucker) [inline src/transformers/modeling_attn_mask_utils.py:202]: the comment was still a bit helpfull IMO

File: src/transformers/modeling_attn_mask_utils.py [hunk 3]
```diff
@@ -199,80 +200,11 @@ def _unmask_unattended(
...
-        else:
-            # Broadcast [left_masked_rows, 1], [left_masked_rows, max_len]
-            mask_slice = (left_masked_rows[:, None], range_tensor)
-
-        expanded_mask[mask_slice] = unmasked_value
+        if expanded_mask.dtype == torch.bool:
+            raise ValueError("AttentionMaskConverter._unmask_unattended expects a float `expanded_mask`, got a BoolTensor.")
 
-        return expanded_mask
+        return expanded_mask.mul(~torch.all(expanded_mask == min_dtype, dim=-1, keepdim=True))
```
Comment (ArthurZucker) [inline src/transformers/modeling_attn_mask_utils.py:207]: ❤️

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks for propagating the changes

File: src/transformers/modeling_attn_mask_utils.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/modeling_attn_mask_utils.py:202]: Agreed! If it's not completely relevant any more can I be updated?

Comment (fxmarty) [reply src/transformers/modeling_attn_mask_utils.py:202]: Yes it should not have been removed, added it back

**Pull request merged**